### DISCUSSION TITLE
Hitsounds are not correctly aligned with the song on non-1x speed rate

### DISCUSSION THREAD

**Discussion by Detze** [ppy/osu#27826] [Ideas] [answered]
### Type

Game behaviour

### Bug description

For the hitsound to play aligned with the song, players must tap early in relation to the song by the amount of their hitsound audio latency (assuming perfectly tuned universal offset). I believe this is why all beatmaps are currently mistimed by on average more or less 23 ms ([as discussed here](https://github.com/ppy/osu/issues/21947#issuecomment-1367998379)), and why [the editor waveform is currently shifted early by 20 ms to compensate](https://github.com/ppy/osu/pull/26136). However, song's audio length is scaled by speed rate changing mods, whereas the hitsound audio latency won't be, and it appears there's no mechanism to correct for that, resulting in hitsounds not being correctly aligned with the song on non-1x speed rate.

To demonstrate: if I play a map with just Auto mod, the hitsounds will be off by 20 ms:

<details>
  <summary>Screenshot</summary>

![usual_1](https://github.com/ppy/osu/assets/92268414/e898c683-ab76-45d1-b47b-421f7c98b5e7)

</details>

However, if I play the same map with Half Time with 0.5x speed and Auto, the hitsounds will be off by 40 ms, noticably more than without Half Time:

<details>
  <summary>Screenshot</summary>

![usual_05](https://github.com/ppy/osu/assets/92268414/34486113-03f9-4873-ba55-294e579dd388)

</details>

And if I play the map with 0.1x (using a forked client with Half Time's `SpeedChange` bindable's `MinValue` changed, for demonstration purposes) the hitsounds will be off by a massive 229 ms:

<details>
  <summary>Screenshot</summary>

![usual_01](https://github.com/ppy/osu/assets/92268414/70def557-229b-48fd-8b91-a31135cdeedd)

</details>

If I correct the beatmap's offset to remove the latency bake-in, then the hitsounds will play at consistent times in relation to the song, regardless of speed rate:
<details>
  <summary>Screenshot</summary>

![fixed_1](https://github.com/ppy/osu/assets/92268414/de80c639-a0a9-465e-802d-d005d30e155e)

</details>

<details>
  <summary>Screenshot</summary>

![fixed_05](https://github.com/ppy/osu/assets/92268414/5820afe5-3c31-4cb4-b75d-14c38b05861c)

</details>

<details>
  <summary>Screenshot</summary>

![fixed_01](https://github.com/ppy/osu/assets/92268414/f78989c8-327c-44d1-b84f-8649f9f4c32a)

</details>

Also note that adjusting universal offset so that the hitsounds are aligned with the song on 1x speed rate does not solve this issue:

<details>
  <summary>Screenshot</summary>

![not_UO](https://github.com/ppy/osu/assets/92268414/a124541c-0ccc-42ce-a263-e8ec06e6a70c)

</details>

Suppose that the average baked-in offset is (negative) 24 ms, and the user's hitsound audio latency is also 24 ms. Then, the relationship between speed rate and hitsound misalignment can be visualized as follows (assuming 0 universal offset and infinitely precise audio): 
| speed rate | hitsound misaligned by |
| --- | --- |
| 0.1x | 216 ms |
| 0.25x | 72 ms |
| 0.5x | 24 ms |
| 0.75x | 8 ms |
| 1x | 0 ms |
| 1.5x | 8 ms |
| 2x | 12 ms |
| 4x | 18 ms | 
| ∞x | 24 ms |
| kx | $\left\lvert 24\left(\frac{1}{k} - 1\right)\right\rvert$ ms |

This is why this issue is most impactful at <1x speed rates, rather than >1x.

The ideal solution would be to remove the latency bake-in in beatmap offsets. A beatmap's timing point times, including beatmap offset, only depend on the song's audio. Audio latency should thus be irrelevant for beatmap offsets. Not sure how feasible that is given all the already existing beatmaps that have been timed with that bake-in. 

Such a solution would make perfect hits's hitsounds play late in relation to the song by {hitsound_audio_latency}, rather than matching the song and requiring the player to tap early by {hitsound_audio_latency}. Achieving the latter behavior would then involve adjusting for hitsound audio latency (and speed rate) elsewhere. If audio latency is non-negligible (I would say it currently is non-negligible, but hopefully the low latency audio system that's currently being developed could reduce it to negligible), the latter might be desirable over the former, as that's the current behavior that players are accustomed to. Due to this audio latency, some players might prefer to play with hitsounds disabled, and so a solution to this issue should also support this use case without having to use a different offset per speed rate.

Note: when using Half Time, the 'Adjust pitch' option should be enabled. If the option is disabled, the song audio is apparently played staggered throughout different times, with silence between the staggered audio portions (screenshot of Half Time with 0.1x speed to see the effect clearly):
<details>
  <summary>Screenshot</summary>

![stagger](https://github.com/ppy/osu/assets/92268414/b526bed8-86d1-4240-93a0-a807cda7d571)

</details>

(note how due to the staggered song portions being played, it appears as though the song is played much earlier than the hitsound, even though the hitsound would normally play about 200 ms before the song)
I don't know whether this is a bug, or some fundamental limitation of playing slowed down audio without adjusting pitch, but while this is in the game, playing Daycore should probably be recommended to players over playing Half Time.

Skin used: argon pro
Beatmap used: [44035 bj.HaLo - Ende (galvenize) [Another]](https://osu.ppy.sh/beatmapsets/44035#osu/148716) (for this beatmap, the difference between the ideal offset without the bake-in and the current offset is around 24 ms)
Version tested: `b559bec`

Seemingly related PR: https://github.com/ppy/osu/pull/27141

### Version

`b559bec`

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/14940971/compressed-logs.zip)

**Comment by peppy** [comment, marked as answer]
The general idea you have here is correct, but the end effect of it is negligible for normal gameplay so it won't be addressed as a priority.

Applying the correction offset at the point of decoding has been discussed in recent times, and can potentially be applied quite easily now that we have a better idea of expectations (via the agreed upon editor visual offset value of 20ms). I'm happy for you or someone else to attempt this, but it should only be someone with a good understanding on how the other pieces of timing need to be adjusted around such a change.

> (note how due to the staggered song portions being played, it appears as though the song is played much earlier than the hitsound, even though the hitsound would normally play about 200 ms before the song)

This is how time stretching algorithms work, it's hard to do much better. And no, I don't see this as a deal-breaker considering we've used it for a decade without issue.
